docs/proposals/2401-llm-trainer-v2/README.md

```diff
@@ -0,0 +1,831 @@
+INFO:torchtune.utils.logging:Learning rate scheduler is initialized.
+1|52|Loss: 2.3697006702423096:   0%|▏                     | 52/25880 [00:24<3:55:01,  1.83it/s]
+```
+
+By adopting `torchtune` as the low-level runtime for LLM fine-tuning, we can easily obtain the flexibility, efficiency and scalability brought by its unique "recipe-config" design, which will surely streamline and scale LLM fine-tuning on Kubernetes.
+
+To hide users from complex Kubernetes configuations, we will provide a simple yet flexible Python SDK wrapping all specifications of models, datasets, training runtime and fine-tuning configs. Like this:
+
+```python
+job_id = TrainingClient().train(
+    dataset_config=HuggingFaceDatasetConfig(
+        storage_uri="tatsu-lab/alpaca",
+    ),
+    trainer=Trainer(
+        fine_tuning_config=TorchTuneConfig(
+            recipe="lora_finetune_single_device",
+            config="llama3_2/1B_lora_single_device",
+            dtype="bf16",
+            batch_size=1,
+            epochs=1,
+            peft_config=LoraConfig(
+                lora_rank=64,
+                lora_alpha=128,
+            ),
+        ),
+        num_nodes=5,
+    ),
+    runtime_ref=llm_runtime,
+)
+```
+
+## Design Details
+
+### Complement `torch` Plugin
+
+**How to Handle Override in Configs**
+
+As is shown in the official guides, we can pass distributed arguments to `torchtune` and override some parameters in the config, like:
+
+```bash
+# Note that tune run only supports launching distributed runs by passing through arguments preceding the recipe directly to torchrun.
+# Official document: https://pytorch.org/torchtune/main/tune_cli.html#run-a-recipe
+# Proof from source code: https://github.com/pytorch/torchtune/blob/75965d4281b9b76c454630d015221b9933c77bf3/torchtune/_cli/run.py#L113-L118
+tune run --nnodes=1 --nproc-per-node=4 lora_finetune_distributed \
+    --config llama3/8B_lora \
+    model.lora_attn_modules=[q_proj,k_proj,v_proj] \
+    model.apply_lora_to_mlp=True \
+    model.lora_rank=64 \
+    model.lora_alpha=128
+
+```
+
+The CLI-based way allows us to mutate these parameters by overriding the `command` and `args` fields in the Trainer Node. By specifying override parameters in the TrainJob `.spec.trainer.command` and `.spec.trainer.args` fields, we can easily mutate these parameters, just like:
+
+```YAML
+apiVersion: kubeflow.org/v2alpha1
+kind: ClusterTrainingRuntime
+metadata:
+  name: torchtune-llm-finetuning
+  labels:
+    training.kubeflow.org/phase: post-training
+spec:
+  mlPolicy:
+    numNodes: 1
+    torch:
+      numProcPerNode: auto
+  template:
+    spec:
+      replicatedJobs:
+        - name: Node
+          template:
+            spec:
+              template:
+                spec:
+                  containers:
+                    - name: trainer
+                      image: <pytorch+cuda+torchtune image>
+                      command:
+                        - tune ls
```

Comment:
Can we wrap the default torchtune config under TrainingRuntime args, and override the user values using `tune run` arguments ?
So the experience will be:
1. ClusterTrainingRuntime contains the default torchtune config to fine-tune model.
2. User can override values using `TorchTuneConfig()` class.